File: backend/app/services/transaction_service.py
from datetime import date as date_type

from sqlalchemy.orm import Session

from app.core.exceptions import (
    BadRequestException,
    NotFoundException,
)
from app.crud import category as category_crud
from app.crud import transaction as transaction_crud
from app.models.transaction import Transaction
from app.schemas.transaction_schema import (
    TransactionCreate,
    TransactionUpdate,
)


# ============================================================
# HELPERS
# ============================================================

def _validate_category_for_user(
    db: Session,
    user_id: int,
    category_id: int,
    expected_type: str,
) -> None:
    """
    Ensure category exists, is visible to user, is active,
    and matches the transaction type.
    """
    category = category_crud.get_category_by_id(db, category_id)

    if not category:
        raise NotFoundException("Category not found")

    # Default categories visible to everyone; custom must belong to user
    if category.user_id is not None and category.user_id != user_id:
        # Return 404 (not 403) to avoid leaking existence of other users' data
        raise NotFoundException("Category not found")

    if not category.is_active:
        raise BadRequestException("Category is inactive")

    if category.type != expected_type:
        raise BadRequestException(
            f"Category type '{category.type}' does not match "
            f"transaction type '{expected_type}'"
        )


# ============================================================
# CREATE
# ============================================================

def create_transaction(
    db: Session,
    user_id: int,
    data: TransactionCreate,
) -> Transaction:
    """Create a transaction after validating the category."""
    _validate_category_for_user(db, user_id, data.category_id, data.type)
    return transaction_crud.create_transaction(db, user_id, data)


# ============================================================
# READ
# ============================================================

def get_transaction_for_user(
    db: Session,
    user_id: int,
    transaction_id: int,
) -> Transaction:
    """
    Get a transaction if it belongs to the user.
    Raises NotFoundException otherwise.
    """
    transaction = transaction_crud.get_transaction_by_id(db, transaction_id)

    if not transaction or transaction.user_id != user_id:
        # 404 not 403 — don't leak existence
        raise NotFoundException("Transaction not found")

    return transaction


def list_user_transactions(
    db: Session,
    user_id: int,
    *,
    type_filter: str | None = None,
    category_id: int | None = None,
    date_from: date_type | None = None,
    date_to: date_type | None = None,
    search: str | None = None,
    sort_by: str = "date",
    sort_order: str = "desc",
    skip: int = 0,
    limit: int = 50,
) -> list[Transaction]:
    """List transactions for a user with optional filters."""
    # Validate date range
    if date_from and date_to and date_from > date_to:
        raise BadRequestException("date_from cannot be after date_to")

    return transaction_crud.list_transactions(
        db,
        user_id,
        type_filter=type_filter,
        category_id=category_id,
        date_from=date_from,
        date_to=date_to,
        search=search,
        sort_by=sort_by,
        sort_order=sort_order,
        skip=skip,
        limit=limit,
    )


# ============================================================
# UPDATE
# ============================================================

def update_transaction(
    db: Session,
    user_id: int,
    transaction_id: int,
    data: TransactionUpdate,
) -> Transaction:
    """
    Update a transaction owned by the user.
    Re-validates the category if category_id or type is changing.
    """
    transaction = get_transaction_for_user(db, user_id, transaction_id)

    # Determine the resulting type (may be updated)
    new_type = data.type if data.type is not None else transaction.type
    new_category_id = (
        data.category_id if data.category_id is not None else transaction.category_id
    )

    # If type or category is changing, re-validate the combination
    if data.type is not None or data.category_id is not None:
        _validate_category_for_user(db, user_id, new_category_id, new_type)

    return transaction_crud.update_transaction(db, transaction, data)


# ============================================================
# DELETE
# ============================================================

def delete_transaction(
    db: Session,
    user_id: int,
    transaction_id: int,
) -> None:
    """Delete a transaction owned by the user."""
    transaction = get_transaction_for_user(db, user_id, transaction_id)
    transaction_crud.delete_transaction(db, transaction)

File: frontend/lib/services/transaction_service.dart
import '../config/api_config.dart';
import '../models/transaction.dart';
import 'api_client.dart';

class TransactionService {
  final ApiClient _apiClient;

  TransactionService({ApiClient? apiClient})
      : _apiClient = apiClient ?? ApiClient();

  /// Fetch transactions with optional filters and pagination.
  Future<List<Transaction>> getTransactions({
    String? type,
    int? categoryId,
    String? dateFrom,
    String? dateTo,
    String? search,
    String? sortBy,
    String? sortOrder,
    int? skip,
    int? limit,
  }) async {
    final queryParams = <String, dynamic>{};
    if (type != null && type.isNotEmpty) queryParams['type'] = type;
    if (categoryId != null) queryParams['category_id'] = categoryId;
    if (dateFrom != null && dateFrom.isNotEmpty) queryParams['date_from'] = dateFrom;
    if (dateTo != null && dateTo.isNotEmpty) queryParams['date_to'] = dateTo;
    if (search != null && search.trim().isNotEmpty) queryParams['search'] = search.trim();
    if (sortBy != null && sortBy.isNotEmpty) queryParams['sort_by'] = sortBy;
    if (sortOrder != null && sortOrder.isNotEmpty) queryParams['sort_order'] = sortOrder;
    if (skip != null) queryParams['skip'] = skip;
    if (limit != null) queryParams['limit'] = limit;

    final response = await _apiClient.get(
      ApiConfig.transactions,
      queryParameters: queryParams.isNotEmpty ? queryParams : null,
    );

    final list = response.data as List<dynamic>;
    return list
        .map((item) => Transaction.fromJson(item as Map<String, dynamic>))
        .toList();
  }

  /// Fetch a single transaction by ID.
  Future<Transaction> getTransaction(int id) async {
    final response = await _apiClient.get(ApiConfig.transaction(id));
    return Transaction.fromJson(response.data as Map<String, dynamic>);
  }

  /// Create a new transaction.
  Future<Transaction> createTransaction({
    required String type,
    required double amount,
    required int categoryId,
    required String date, // 'YYYY-MM-DD'
    String? note,
  }) async {
    final response = await _apiClient.post(
      ApiConfig.transactions,
      data: {
        'type': type,
        'amount': amount,
        'category_id': categoryId,
        'date': date,
        if (note != null && note.trim().isNotEmpty) 'note': note.trim(),
      },
    );

    return Transaction.fromJson(response.data as Map<String, dynamic>);
  }

  /// Update an existing transaction.
  Future<Transaction> updateTransaction(
    int id, {
    String? type,
    double? amount,
    int? categoryId,
    String? date,
    String? note,
  }) async {
    final data = <String, dynamic>{};
    if (type != null) data['type'] = type;
    if (amount != null) data['amount'] = amount;
    if (categoryId != null) data['category_id'] = categoryId;
    if (date != null) data['date'] = date;
    if (note != null) data['note'] = note.trim();

    final response = await _apiClient.put(
      ApiConfig.transaction(id),
      data: data,
    );

    return Transaction.fromJson(response.data as Map<String, dynamic>);
  }

  /// Delete a transaction by ID.
  Future<void> deleteTransaction(int id) async {
    await _apiClient.delete(ApiConfig.transaction(id));
  }
}
